bedrock/3_EntityWorkflow/workflows.py

```diff
@@ -0,0 +1,174 @@
+        self.continue_as_new_per_turns: int = 6
+        self.chat_ended: bool = False
+
+    @workflow.run
+    async def run(
+        self,
+        params: BedrockParams = None,
+    ) -> str:
+
+        if params and params.conversation_summary:
+            self.conversation_history.append(
+                ("conversation_summary", params.conversation_summary)
+            )
+
+        if params and params.prompt_queue:
+            for prompt in params.prompt_queue:
+                self.prompt_queue.append(prompt)
+
+        summary_activity_task: Optional[asyncio.Task] = None
+
+        while True:
+            workflow.logger.info("Waiting for prompts...")
+
+            # Wait for a chat message (signal) or timeout
+            await workflow.wait_condition(
+                lambda: bool(self.prompt_queue) or self.chat_ended
+            )
+
+            # if end chat signal was sent
+            if self.chat_ended:
+                # the workflow might be continued as new without any
+                # chat to summarize
+                if summary_activity_task is not None:
+                    # ensure conversation summary task has finished
+                    # before closing the workflow (avoid race)
+                    await workflow.wait_condition(
+                        lambda: summary_activity_task.done()
+                    )
+                else:
+                    # conversation history from previous workflow
+                    self.conversation_summary = params.conversation_summary
+
+                workflow.logger.info(
+                    "Chat ended. Conversation summary:\n" +
+                    f"{self.conversation_summary}"
+                )
+
+                return f"{self.conversation_history}"
+
+            # Fetch next user prompt and add to conversation history
+            prompt = self.prompt_queue.popleft()
+            self.conversation_history.append(("user", prompt))
+
+            workflow.logger.info("Prompt: " + prompt)
+
+            # Send prompt to Amazon Bedrock
+            response = await workflow.execute_activity(
+                prompt_bedrock,
+                self.prompt_with_history(prompt),
+                schedule_to_close_timeout=timedelta(seconds=20),
+            )
+
+            workflow.logger.info(f"{response}")
+
+            # Append the response to the conversation history
+            self.conversation_history.append(("response", response))
+
+            # summarize the conversation to date using Amazon Bedrock
+            # uses start_activity with a callback
+            # so it doesn't block new messages being sent to Amazon Bedrock
+            summary_activity_task = workflow.start_activity(
+                prompt_bedrock,
+                self.prompt_summary_from_history(),
+                schedule_to_close_timeout=timedelta(seconds=20),
+            )
+            summary_activity_task.add_done_callback(self.summary_complete)
```

Comment:
This is racy. What if a summary task from an early iteration of the loop finishes after a summary task from a later iteration? You should make summary occur at the end IMO or on user request.

But if you do want to leave it like this, you should make sure only one runs at a time and it'd probably be clearer to use an `async def` to do it and store the result instead of a done callback (and use `asyncio.create_task` to run it in the background if needed). This approach does not properly account for failures of the activity (will throw in the callback which goes nowhere).

Reply:
This is a good point, thanks Chad. I might refactor this so the summary is triggered only when needed (before CAN / and at the end).